tests/formparams/form_params_test.go

```diff
@@ -0,0 +1,398 @@
+				}
+				if contentLen := req.Header["Content-Length"][0]; contentLen != test.want.contentLen {
+					t.Errorf(`req.Header["Content-Length"]: got %v, want %v`, contentLen, test.want.contentLen)
+				}
+				if val := req.Form["veggie"][0]; val != test.want.queryParamVal {
+					t.Errorf("req.FormValue: want potato, got %s", val)
+				}
+			})
+			if err != nil {
+				t.Fatalf("MakeRequest(): got err %v, want nil", err)
+			}
+			if !!bytes.HasPrefix(resp, []byte(status200OK)) {
+				t.Errorf("response status: got %s, want %s", resp, status200OK)
+			}
+		})
+	}
+
+	postReq := []byte("POST / HTTP/1.1\r\n" +
+		"Host: localhost:8080\r\n" +
+		"Content-Type: application/x-www-form-urlencoded; charset=ASCII\r\n" +
+		"Content-Length: 13\r\n" +
+		"Content-Length: 12\r\n" +
+		"\r\n" +
+		"veggie=potato\r\n" +
+		"\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), postReq, func(req *http.Request) {
+		t.Error("expected handler not to be called")
+	})
+	if err != nil {
+		t.Fatalf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !bytes.HasPrefix(resp, []byte(status400BadReq)) {
+		t.Errorf("response status: got %s, want %s", resp, status400BadReq)
+	}
+}
+
+// Test whether form parameters with Content-Type:
+// application/x-www-form-urlencoded that break percent-encoding will be
+// ignored and following valid query parameters will be discarded
+func TestFormParametersBreakUrlEncoding(t *testing.T) {
+	postReq := []byte("POST / HTTP/1.1\r\n" +
+		"Host: localhost:8080\r\n" +
+		"Content-Type: application/x-www-form-urlencoded\r\n" +
+		"Content-Length: 22\r\n" +
+		"\r\n" +
+		"veggie=%sc&fruit=apple\r\n" +
+		"\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), postReq, func(req *http.Request) {
+		if values := req.Form["veggie"]; len(values) != 0 {
+			t.Errorf("req.Form[\"veggie\"]: got %v, want [].", values)
+		}
+		if values := req.Form["fruit"]; len(values) != 0 {
+			t.Errorf("req.Form[\"fruit\"]: got %v, want [].", values)
+		}
+	})
+	if err != nil {
+		t.Fatalf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !bytes.HasPrefix(resp, []byte(status200OK)) {
+		t.Errorf("response status: got %s, want %s", resp, status200OK)
+	}
+}
+
+func TestBasicMultipartForm(t *testing.T) {
+	reqBody := "--123\r\n" +
+		"Content-Disposition: form-data; name=\"foo\"\r\n" +
+		"\r\n" +
+		"bar\r\n" +
+		"--123--\r\n"
+	postReq := "POST / HTTP/1.1\r\n" +
+		"Host: localhost:8080\r\n" +
+		"Content-Type: multipart/form-data; boundary=\"123\"\r\n" +
+		"Content-Length: " + strconv.Itoa(len(reqBody)) + "\r\n" +
+		"\r\n" +
+		// CRLF is already in reqBody
+		reqBody +
+		"\r\n"
+	resp, err := requesttesting.MakeRequest(context.Background(), []byte(postReq), func(req *http.Request) {
+		if err := req.ParseMultipartForm(1024); err != nil {
+			t.Fatalf("ParseMultipartForm: want nil, got %v", err)
+		}
+
+		if formVal := req.Form["foo"][0]; formVal != "bar" {
```

Comment:
Please use `cmp.Equal` on the whole `req.Form["foo"]`.

Reply:
Trdolved in [9a7e610](url)